File: Algorithm/053. Maximum Subarray.py
# -*- coding: utf-8 -*-
# @Author: WuLC
# @Date:   2016-05-03 14:49:45
# @Last modified by:   WuLC
# @Last Modified time: 2016-05-03 14:54:07
# @Email: liangchaowu5@gmail.com

"""
动态规划，因为只需要求大小，所以不用新开数组，直接在原数组操作
经过DP后nums[i]的大小为以原来nums[i]作为子串最后的一个元素的最大总和
"""
class Solution(object):
    def maxSubArray(self, nums):
        """
        :type nums: List[int]
        :rtype: int
        """
        n = len(nums)
        for i in xrange(1,n):
            if nums[i-1]>0:
                nums[i] = nums[i]+nums[i-1]
                
        max_num = nums[0]
        for i in xrange(1,n):
            if nums[i]>max_num:
                max_num = nums[i]
        
        return max_num
            

File: Algorithm/Java/053. Maximum Subarray.java
/**
* Author: LC
* Date:   2016-12-26 22:06:04
* Last modified by:   WuLC
* Last Modified time: 2016-12-26 22:06:16
* Email: liangchaowu5@gmail.com
*/


// dp
public class Solution 
{
    public int maxSubArray(int[] nums) 
    {
        if (nums.length == 0) return 0;
        int result = nums[0], previousMax = nums[0], currMax;
        for(int i = 1;i < nums.length; i++)
        {
            currMax = Math.max(nums[i], nums[i]+previousMax);
            result = Math.max(currMax, result);
            previousMax = currMax;
        }
        return result;
    }
}